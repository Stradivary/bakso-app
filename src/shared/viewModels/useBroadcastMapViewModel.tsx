import { Stack, Text } from "@mantine/core";
import { useDisclosure } from "@mantine/hooks";
import { modals } from "@mantine/modals";
import { notifications as notify } from "@mantine/notifications";
import L, { LatLng } from "leaflet";
import React, { useCallback, useMemo, useState } from "react";
import { useAuth } from "../hooks/useAuth";
import { useTracker } from "../hooks/useTracker";
import { User } from "../models/BroadcastMapModel";
import { calculateDistance } from "../services/trackerServices";

export const useBroadcastMapViewModel = (
  location: {
    latitude: number;
    longitude: number;
  } | null,
) => {
  const { session, logout } = useAuth();
  const userId = session?.user?.id as string;
  const userName = session?.user?.user_metadata?.display_name;
  const userRole = session?.user?.user_metadata?.role as
    | "seller"
    | "buyer";
  const userLocation = [session?.user?.user_metadata?.latitude, session?.user?.user_metadata?.longitude];
  const [exitModalOpened, { close: exitModalClose, open: openModal }] =
    useDisclosure();
  const [lastValidLatLng, setLastValidLatLng] = useState<LatLng | null>(null);

  const fixedLocation = useMemo<LatLng>(() => {
    if (userRole === 'buyer') {
      return new L.LatLng(userLocation[0], userLocation[1]);
    }

    if (!location?.latitude || !location?.longitude && lastValidLatLng) {
      return lastValidLatLng ?? new L.LatLng(0, 0);
    }

    const LatLng = new L.LatLng(location?.latitude, location?.longitude);

    setLastValidLatLng(LatLng);

    return LatLng;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location?.latitude, location?.longitude]);

  const {
    nearbyUsers,
    sendPing,
    notifications,
    handleLocationUpdate,
    deactivateUser,
  } = useTracker(
    userId,
    userRole,
    fixedLocation,
    session?.user?.user_metadata?.name ?? "",
  );

  const [mapRef, setMapRef] = React.useState<L.Map | null>(null);

  const handleRecenter = React.useCallback(() => {
    if (mapRef && location) {
      const newLocation = [location?.latitude ?? 0, location?.longitude ?? 0];
      mapRef.setView(
        new L.LatLng(newLocation[0], newLocation[1]),
        mapRef.getZoom(),
        {
          animate: true,
          duration: 1,
        },
      );
    }
  }, [mapRef, location]);

  const handleExit = React.useCallback(() => {
    deactivateUser();
    logout();
    exitModalClose();
  }, [deactivateUser, exitModalClose, logout]);

  const handleMarkerClick = useCallback(
    (nearbyUser: User) => {
      if (nearbyUser.role === "seller") {
        const buyerLocation = new L.LatLng(
          location?.latitude ?? 0,
          location?.longitude ?? 0,
        );
        const sellerLocation = new L.LatLng(
          nearbyUser.location.lat,
          nearbyUser.location.lng,
        );
        const distanceInMeters = calculateDistance(
          buyerLocation,
          sellerLocation,
        );
        const walkingSpeed = 1.4; // meters per second
        const estimatedTimeInMinutes = Math.round(
          distanceInMeters / walkingSpeed / 60,
        );

        modals.openConfirmModal({
          title: "Pesan Bakso",
          children: (
            <Stack gap={4}>
              <Text size="sm">
                üçú Apakah Anda ingin memesan bakso dari {nearbyUser.userName}?
              </Text>
              <Text size="xs" c="dimmed">
                Estimasi waktu: {estimatedTimeInMinutes} menit berjalan kaki
              </Text>
            </Stack>
          ),
          labels: { cancel: "Batal", confirm: "Pesan" },
          onConfirm: () => {
            sendPing(nearbyUser.user_id);
            notify.show({
              title: "Pesan Bakso",
              message: "Pesan bakso telah dikirim",
              color: "blue",
            });
          },
        });
      }
    },
    [location, sendPing],
  );

  const centerLocation = useMemo(
    () => {

      if (userRole === "buyer" && fixedLocation)
        return fixedLocation;

      return new L.LatLng(location?.latitude ?? 0, location?.longitude ?? 0);

    },
    [location],
  );

  return {
    userId,
    userName,
    session,
    userRole,
    location,
    nearbyUsers,
    handleLocationUpdate,
    notifications,
    exitModalOpened,
    openModal,
    exitModalClose,
    handleRecenter,
    handleExit,
    handleMarkerClick,
    setMapRef,
    fixedLocation,
    centerLocation,
  };
};
