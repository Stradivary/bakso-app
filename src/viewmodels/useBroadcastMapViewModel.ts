import { User } from "@/models/BroadcastMapModel";
import { calculateDistance } from "@/viewmodels/hooks/calculateDistance";
import { useTracker } from "@/viewmodels/hooks/useTracker";
import { useAuth } from "@/views/providers/authProvider";
import { useDisclosure } from "@mantine/hooks";
import L, { LatLng } from "leaflet";
import React, { useCallback, useMemo, useState } from "react";

export const useBroadcastMapViewModel = (
  location: {
    latitude: number;
    longitude: number;
  } | null,
) => {
  const { session, logout } = useAuth();
  const userId = session?.user?.id as string;
  const userName = session?.user?.user_metadata?.display_name;
  const userRole = session?.user?.user_metadata?.role as "seller" | "buyer";
  const userLocation = [
    session?.user?.user_metadata?.latitude,
    session?.user?.user_metadata?.longitude,
  ];
  const [exitModalOpened, { close: exitModalClose, open: openModal }] =
    useDisclosure();
  const [lastValidLatLng, setLastValidLatLng] = useState<LatLng | null>(null);

  const fixedLocation = useMemo<LatLng>(() => {
    if (userRole === "buyer") {
      return new L.LatLng(userLocation[0], userLocation[1]);
    }

    if (!location?.latitude || (!location?.longitude && lastValidLatLng)) {
      return lastValidLatLng ?? new L.LatLng(0, 0);
    }

    const LatLng = new L.LatLng(location?.latitude, location?.longitude);

    setLastValidLatLng(LatLng);

    return LatLng;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location?.latitude, location?.longitude]);

  const {
    nearbyUsers,
    sendPing,
    notifications,
    handleLocationUpdate,
    deactivateUser,
  } = useTracker(
    userId,
    userRole,
    fixedLocation,
    session?.user?.user_metadata?.name ?? "",
  );

  const [mapRef, setMapRef] = React.useState<L.Map | null>(null);

  const handleRecenter = React.useCallback(() => {
    if (mapRef && location) {
      const newLocation = [location?.latitude ?? 0, location?.longitude ?? 0];
      mapRef.setView(
        new L.LatLng(newLocation[0], newLocation[1]),
        mapRef.getZoom(),
        {
          animate: true,
          duration: 1,
        },
      );
    }
  }, [mapRef, location]);

  const handleExit = React.useCallback(() => {
    deactivateUser();
    logout();
    exitModalClose();
  }, [deactivateUser, exitModalClose, logout]);

  const handleMarkerClick = useCallback(
    (nearbyUser: User) => {
      if (nearbyUser.role === "seller") {
        const buyerLocation = new L.LatLng(
          location?.latitude ?? 0,
          location?.longitude ?? 0,
        );
        const sellerLocation = new L.LatLng(
          nearbyUser.location.lat,
          nearbyUser.location.lng,
        );
        const distanceInMeters = calculateDistance(
          buyerLocation,
          sellerLocation,
        );
        const walkingSpeed = 1.4; // meters per second
        const estimatedTimeInMinutes = Math.round(
          distanceInMeters / walkingSpeed / 60,
        );

        return {
          nearbyUser,
          estimatedTimeInMinutes,
        };
      }
    },
    [location],
  );

  const centerLocation = useMemo(() => {
    if (userRole === "buyer" && fixedLocation) return fixedLocation;

    return new L.LatLng(location?.latitude ?? 0, location?.longitude ?? 0);
  }, [location]);

  return {
    userId,
    userName,
    session,
    userRole,
    location,
    nearbyUsers,
    handleLocationUpdate,
    notifications,
    exitModalOpened,
    openModal,
    exitModalClose,
    handleRecenter,
    handleExit,
    handleMarkerClick,
    setMapRef,
    fixedLocation,
    centerLocation,
    sendPing,
  };
};
