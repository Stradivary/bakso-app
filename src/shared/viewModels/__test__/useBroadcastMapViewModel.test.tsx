import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { renderHook, act } from "@testing-library/react";
import { useBroadcastMapViewModel } from "../useBroadcastMapViewModel";
import { modals } from "@mantine/modals";

vi.mock("../../hooks/useAuth", () => ({
  useAuth: () => ({
    session: {
      user: {
        id: "test-id",
        user_metadata: { name: "Test User" },
      },
    },
    logout: vi.fn(),
  }),
}));

vi.mock("../../hooks/useLocation", () => ({
  useLocation: () => ({
    location: { latitude: 0, longitude: 0 },
  }),
}));

vi.mock("../hooks/useTracker", () => ({
  useTracker: () => ({
    nearbyUsers: [],
    handleLocationUpdate: vi.fn(),
    handlePing: vi.fn(),
    notifications: [],
    deactivateUser: vi.fn(),
  }),
}));

vi.mock("@mantine/modals", () => ({
  modals: {
    openConfirmModal: vi.fn(),
  },
}));

describe("useBroadcastMapViewModel", () => {
  beforeEach(() => {
    vi.stubGlobal("sessionStorage", {
      getItem: () => "buyer",
      removeItem: vi.fn(),
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("should initialize with correct values", () => {
    const { result } = renderHook(() => useBroadcastMapViewModel());

    expect(result.current.userId).toBe("test-id");
    expect(result.current.userRole).toBe("buyer");
  });

  it("should handle exit correctly", () => {
    const { result } = renderHook(() => useBroadcastMapViewModel());

    act(() => {
      result.current.handleExit();
    });

    expect(result.current.handleExit).toBeDefined();
  });

  it("should handle map recenter", () => {
    const { result } = renderHook(() => useBroadcastMapViewModel());
    const mockMap = {
      setView: vi.fn(),
      getZoom: vi.fn().mockReturnValue(15),
    };

    act(() => {
      result.current.setMapRef(mockMap as unknown as L.Map);
      result.current.handleRecenter();
    });

    expect(result.current.setMapRef).toBeDefined();
  });

  it("should handle marker click for seller", () => {
    const { result } = renderHook(() => useBroadcastMapViewModel());
    const mockNearbyUser = {
      role: "seller" as const,
      user_id: "seller-id",
      userName: "Seller",
      location: { lat: 1, lng: 1 },
    };

    act(() => {
      result.current.handleMarkerClick(
        mockNearbyUser,
        { latitude: 0, longitude: 0 },
        vi.fn(),
      )?.({} as L.LeafletMouseEvent);
    });

    expect(modals.openConfirmModal).toHaveBeenCalled();
  });
});
