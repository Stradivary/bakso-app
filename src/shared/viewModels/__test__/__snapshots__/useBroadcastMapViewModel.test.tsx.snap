// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useBroadcastMapViewModel > handles exit modal actions 1`] = `
{
  "current": {
    "centerLocation": {
      "lat": 0,
      "lng": 0,
    },
    "exitModalClose": [Function],
    "exitModalOpened": false,
    "handleExit": [Function],
    "handleLocationUpdate": [MockFunction spy],
    "handleMarkerClick": [Function],
    "handleRecenter": [Function],
    "location": {
      "latitude": 0,
      "longitude": 0,
    },
    "nearbyUsers": [],
    "notifications": [],
    "openModal": [Function],
    "session": {
      "access_token": "",
      "expires_in": 0,
      "refresh_token": "",
      "token_type": "",
      "user": {
        "app_metadata": {},
        "aud": "",
        "created_at": "",
        "id": "test-id",
        "user_metadata": {
          "name": "Test User",
        },
      },
    },
    "setMapRef": [Function],
    "userId": "test-id",
    "userName": "Test User",
    "userRole": "buyer",
  },
}
`;
